### ISSUE TITLE
Use local time of Home Assistant rather than browser.

### ISSUE THREAD

**Issue by dunxd** [pkissling/clock-weather-card#157]
**Is your feature request related to a problem? Please describe.**
I'm on holiday in a different time zone and notice the time displayed is where I am, not where my Home Assistant is. I have plenty of ways to find out what time it is for me, but when looking at my HA it is more useful to know what time it is _there_.

I've confirmed this by changing the timezone on my phone, and see the time change in the card.

**Describe the solution you'd like**
Show the time local to the HA install. This could be by getting the time from HA itself or setting the timezone to use in the config. Timezone would allow flexibility for those that want to use HA as a clock.

**Comment by miholobolo** [comment]
I would love to have this as well. In my case i use the clock on a screen device that is cut from the internet, and has no reliable internal clock. I was hoping that could use the time of the HA.

**Comment by pkissling** [comment]
thanks for the request.
when i initially implemented the card, i realised that `sensor.time` is optional in home assistant and must be [configured explicilty](https://www.home-assistant.io/integrations/time_date/)
hence i could not rely on the sensor being available in every HA instance. thus i decided to get the time from the current browser.

what i could do though, is to allow providing an optional property `time_sensor: sensor.time` and read the current time from there. but since `sensor.time` only holds the time, i would additionally need `sensor.date` to show the weekday and the corresponding date, which makes the configuration a bit clunky since a user would have to configure both values `time_sensor` and `date_sensor` to make it work.

**Comment by Progaros** [comment]
I use LibreWolf Browser which only uses GMT 0. I would really like an an option to define a custom offset or timezone

([get browser time zone](https://stackoverflow.com/questions/1091372/getting-the-clients-time-zone-and-offset-in-javascript), calculate difference to user setting and set new time)

**Comment by Ih8rain2** [comment]
Can’t you use the time zone setting as set in HA?  I have the same issue since I work on a different continent from where my house is.

**Comment by pkissling** [comment]
> Can’t you use the time zone setting as set in HA? I have the same issue since I work on a different continent from where my house is.

thanks for the hint. that might be feasible and might make the configuration/implementation easier.
i'll have a look into it in the next couple of days

**Comment by rad2000** [comment]
If it doesn't add too much complexity, I'd love the ability to specify any sensor for the date/time. I have a wall panel where I'd like to display the time and weather in different cities where I have family. I can currently do this with the weather since I can specify any sensor, but I can't for the time.

**Comment by dunxd** [comment]
I don't think it needs to use timezone settings - just get the time from the local server instance instead of the client browser. Being able to specify the sensor to use adds flexibility as described. A default that uses the browser time if no sensor is specified would be ok, since HA doesn't enable `sensor.time` by default.

**Comment by dunxd** [comment]
Much prefer the suggestion to use any sensor - this allows more flexibility. Is there a way to have a default that can be overridden?

**Comment by pedromfa** [comment]
@pkissling any development on this? Would really appreciate to have the host machine local time option. :) Thank you so much for this card!

**Comment by Progaros** [comment]
i currently use this userscript with e.g. violentmonkey:

change `YOUR_DOMAIN` and `germanTimeZoneOffset` 
```
// ==UserScript==
// @name         Deutsche Zeitzone
// @namespace   Violentmonkey Scripts
// @match       https://YOUR_DOMAIN
// @version      0.1
// @description  Überschreibt das Date-Objekt, um die deutsche Zeitzone zu verwenden
// @author       Me
// @grant        none
// @run-at       document-start
// ==/UserScript==

(function() {
    'use strict';

    const today = new Date();
    const month = today.getMonth() + 1; // Januar ist 0, daher +1
    const day = today.getDate();
    const hour = today.getHours();

    let isSummerTime;

    if (month > 3 && month < 10) { // Sommerzeit gilt von April bis September
      isSummerTime = true;
    } else if (month === 3 && day >= 29 && hour >= 2) { // letzter Sonntag im März um 2 Uhr
      isSummerTime = true;
    } else if (month === 10 && day <= 25 && hour < 3) { // letzter Sonntag im Oktober um 3 Uhr
      isSummerTime = true;
    } else {
      isSummerTime = false;
    }

    const germanTimeZoneOffset = isSummerTime ? -120 : -60; // Deutsche Zeitzone ist UTC+1 (+2 während Sommerzeit)
    const originalDate = Date;

    function getGermanDate(...args) {
        if (args.length === 0) {
            const now = new originalDate();
            return new originalDate(now.getTime() + (now.getTimezoneOffset() - germanTimeZoneOffset) * 60000);
        } else {
            return new originalDate(...args);
        }
    }

    // Überschreibe das Date-Objekt
    const newDate = function(...args) {
        return getGermanDate(...args);
    };

    // Kopiere die Prototypen und statischen Methoden
    Object.setPrototypeOf(newDate, originalDate);
    Object.setPrototypeOf(newDate.prototype, originalDate.prototype);
    for (const prop of Object.getOwnPropertyNames(originalDate)) {
        if (typeof originalDate[prop] === 'function') {
            newDate[prop] = originalDate[prop].bind(originalDate);
        } else {
            Object.defineProperty(newDate, prop, Object.getOwnPropertyDescriptor(originalDate, prop));
        }
    }

    // Date-Objekt überschreiben
    Date = newDate;
    window.Date = newDate;
})();
```

**Comment by pkissling** [comment]
> @pkissling any development on this? Would really appreciate to have the host machine local time option. :) Thank you so much for this card!

i did the code changes today (introducing an attribute `use_browser_time` which can be set to `false` to use time from `time_zone` configured in HA.
i will have to do some testing tomorrow and will afterwards release a new version

**Comment by pedromfa** [comment]
@pkissling thank you so much! 😎

**Comment by pedromfa** [comment]
![Screenshot_2023-04-24-08-39-09-448-edit_io homeassistant companion android](https://user-images.githubusercontent.com/47584375/233880826-2aa9b146-1078-403e-b1eb-d2c8a3e2de77.jpg)
Hi again. Now it seems there's a 12h difference between Ha time and the one your card displays. See screenshot attached. Perhaps some error in the background? Appreciated the efforts here @pkissling. 😎😉

**Comment by teddy-ptmedia** [comment]
Is it 12hr or is one showing 12hr and one showing 24hr formats?On Apr 23, 2023, at 3:44 PM, pedromfa ***@***.***> wrote:﻿

Hi again. Now it seems there's a 12h difference between Ha time and the one your card displays. See screenshot attached. Perhaps some error in the background? Appreciated the efforts here @pkissling. 😎😉

**Comment by pedromfa** [comment]
Both time format gives me the wrong time by +12h

**Comment by pkissling** [comment]
> Both time format gives me the wrong time by +12h

Can you please give additional information?
Your current time zone (browser), your current time, the time zone configured in HA and the browser you are using. i want to try to reproduce the problem

**Comment by pedromfa** [comment]
My current time zone is UTC+6 (12:42lt). My HA host is UTC+1(0642lt). And I'm using the android companion app to access my instance. Sorry for all this trouble. 😁

**Comment by pkissling** [comment]
> 

What time zone is configured in your HA instance? e.g. `time_zone: Europe/Berlin`

**Comment by pedromfa** [comment]
![Screenshot_2023-04-24-16-17-13-942_io homeassistant companion android](https://user-images.githubusercontent.com/47584375/233954697-a0b93031-9b3c-4a39-9280-aee300ec5950.jpg)

**Comment by dunxd** [comment]
Just updated after getting back from my holiday, but tested by changing the timezone on my phone.  Works as expected which is awesome.  Am I right in thinking you achieved this by applying the timezone for HA to the time coming from the browser?

IMHO `use_browser_time` should be false by `default` as it should correspond to the location that weather is being shown for in the same card!

**Comment by pkissling** [comment]
> Just updated after getting back from my holiday, but tested by changing the timezone on my phone. Works as expected which is awesome. Am I right in thinking you achieved this by applying the timezone for HA to the time coming from the browser?
> 
> IMHO `use_browser_time` should be false by `default` as it should correspond to the location that weather is being shown for in the same card!

yes, agreed. i have this on the radar for the next major version of the card. by default the card would then fetch the current time from HA instead of the browser and a user would be able to disable it by setting `use_browser_time: false`.

the reason i didn't do this in the current release is because i did not want to break existing dashboards by changing the default logic of where the card gets the time from
https://github.com/pkissling/clock-weather-card/milestone/3

**Comment by pkissling** [comment]
> ![Screenshot_2023-04-24-16-17-13-942_io homeassistant companion android](https://user-images.githubusercontent.com/47584375/233954697-a0b93031-9b3c-4a39-9280-aee300ec5950.jpg)

thanks for the feedback. i released 1.0.14 today. hope its fixed now. otherwise please open a new issue

**Comment by pedromfa** [comment]
@pkissling it's wonderful after last update! Many many thanks for this! 😎
